sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp

```diff
@@ -559,15 +559,39 @@ lsc_slm_block_load(uint32_t offset, __ESIMD_NS::simd_mask<1> pred = 1) {
   constexpr uint16_t AddressScale = 1;
   constexpr int ImmOffset = 0;
   constexpr lsc_data_size FDS = detail::finalize_data_size<T, DS>();
-  static_assert(FDS == lsc_data_size::u32 || FDS == lsc_data_size::u64,
-                "Transposed load is supported only for data size u32 or u64");
-  constexpr detail::lsc_vector_size VS = detail::to_lsc_vector_size<NElts>();
+
+  constexpr int SmallIntFactor =
+      (FDS == lsc_data_size::u16) ? 2 : (FDS == lsc_data_size::u8 ? 4 : 1);
+  static_assert(NElts > 0 && NElts % SmallIntFactor == 0,
+                "Number of elements is not supported by Transposed load");
+
+  constexpr int FactoredNElts = NElts / SmallIntFactor;
+  detail::check_lsc_vector_size<FactoredNElts>();
+
+  // Prepare template arguments for the call of intrinsic.
+  constexpr lsc_data_size ActualDS = sizeof(T) == 8
+                                         ? __ESIMD_ENS::lsc_data_size::u64
+                                         : __ESIMD_ENS::lsc_data_size::u32;
+
+  using LoadElemT = __ESIMD_DNS::__raw_t<
```

Comment:
This most likely works wrong if T is tfloat32 as it will bring us to L585 that doesn't have bit-casts.